### ISSUE TITLE
Deadlock while serving a REST-Endpoint after upgrading from 20.0.0.1

### ISSUE THREAD

**Issue by c-koell** [OpenLiberty/open-liberty#12704]
**Describe the bug**
We have a REST-Endpoint and after upgrading from 20.0.0.1 to all newer versions we see following deadlock after a while running the application.

[ThreadDump.txt](https://github.com/OpenLiberty/open-liberty/files/4802569/ThreadDump.txt)

**Steps to Reproduce**
Nothing special ...

**Expected behavior**
No Deadlock ..

**Diagnostic information:**
 - OpenLiberty Version: [e.g. 20.0.0.2] and all newer
 - Java Version: 
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.18.0, JRE 11 Windows Server 2019 amd64-64-Bit Compressed References 20200116_435 (JIT enabled, AOT enabled)
OpenJ9   - 6968c18d7
OMR      - 7a1b0239a
JCL      - 104cab2452 based on jdk-11.0.6+10)
 - server.xml configuration
```<?xml version="1.0" encoding="UTF-8"?>
<server>
	<featureManager>
		<feature>mpConfig-1.3</feature>
		<feature>localConnector-1.0</feature>
		<feature>webProfile-8.0</feature>
		<feature>mpFaultTolerance-1.1</feature>
		<feature>javaMail-1.6</feature>		
	</featureManager>

	<httpEndpoint host="*" httpPort="${server.httpPort}" httpsPort="${server.httpsPort}" id="defaultHttpEndpoint" />
	<applicationManager autoExpand="true" />
	<httpSession invalidationTimeout="10m" useContextRootAsCookiePath="true" cookieSecure="true"/>
</server>
```

**Comment by c-koell** [comment]
I think the deadlock occurs in the following code
```
"Default Executor-thread-107" Id=263 in BLOCKED on lock=java.lang.StringBuffer@c32fef9
     owned by oak-scheduled-executor-59 Id=366
    at 11.0.6/java.lang.StringBuffer.setLength(StringBuffer.java:232)
      - locked java.lang.StringBuffer@c32fef9
    at org.apache.cxf.jaxrs.interceptor.CachedTime.updateTime(CachedTime.java:86)
    at org.apache.cxf.jaxrs.interceptor.CachedTime.getTimeAsString(CachedTime.java:134)
      - locked org.apache.cxf.jaxrs.interceptor.CachedTime@a49e679a
    at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.setResponseDate(JAXRSOutInterceptor.java:448)
```
```
"Default Executor-thread-60" Id=188 in BLOCKED on lock=org.apache.cxf.jaxrs.interceptor.CachedTime@a49e679a
     owned by Default Executor-thread-107 Id=263
    at org.apache.cxf.jaxrs.interceptor.CachedTime.getTimeAsString(CachedTime.java:134)
      - locked org.apache.cxf.jaxrs.interceptor.CachedTime@a49e679a
 ```

**Comment by jhanders34** [comment]
Currently in the thread dump you provided, I do not see any deadlock.  If you could get a J9 thread dump frequently called a javacore, that would be more helpful.  In order to get a javacore with a J9 JDK, you would simply calling wlp/bin/server javadump <serverName>

**Comment by c-koell** [comment]
You are right no Thread is marked as a deadlock.
But as you can see the Thread "Default Executor-thread-107" is blocked and the Request hangs.
All new Request's to the REST Endpoint hangs and resulting in a blocked thread like "Default Executor-thread-60" and "Default Executor-thread-170".

I will create a java dump as you have describe and attach the output.

**Comment by c-koell** [comment]
I have tested follwing command
`wlp/bin/server javadump <serverName>` under windows but i always got a warning that the javadump could not be generated.
In the messages log of the server is following log
`[23.06.20, 10:17:49:904 MESZ] 0000001b com.ibm.ws.kernel.launch.internal.FrameworkManager           A CWWKE0067I: Es wurde eine Java-Speicherauszugsanforderung empfangen.
`
The console log contains following log
`[AUDIT   ] CWWKE0067I: Es wurde eine Java-Speicherauszugsanforderung empfangen.
JVMJ9VM079W triggerOneOffDump(java) erfordert j9dmp29
`
The Server is installed on windows as a windows service.

**Comment by c-koell** [comment]
Oh sorry i have found that i set following jvm option
-Xdump:none
I will remove it and try again.

**Comment by c-koell** [comment]
I have attached 4 Javacore Files taken during the deadlock ...
[javacore.20200623.132300.6820.0002.txt](https://github.com/OpenLiberty/open-liberty/files/4819018/javacore.20200623.132300.6820.0002.txt)
[javacore.20200623.132305.6820.0003.txt](https://github.com/OpenLiberty/open-liberty/files/4819019/javacore.20200623.132305.6820.0003.txt)
[javacore.20200623.132311.6820.0004.txt](https://github.com/OpenLiberty/open-liberty/files/4819020/javacore.20200623.132311.6820.0004.txt)
[javacore.20200623.132315.6820.0005.txt](https://github.com/OpenLiberty/open-liberty/files/4819021/javacore.20200623.132315.6820.0005.txt)

**Comment by jhanders34** [comment]
Thank you very much.  This helps a lot.  The thread dump function (javacore) in J9 JVMs gives some nice details like:

2LKDEADLOCKTHR  Thread "Default Executor-thread-1733" (0x0000000002149400)
3LKDEADLOCKWTR    is waiting for:
4LKDEADLOCKMON      sys_mon_t:0x00000288FB2B1E48 infl_mon_t: 0x00000288FB2B1EC8:
4LKDEADLOCKOBJ      java/lang/StringBuffer@0x00000000C6075228
3LKDEADLOCKOWN    which is owned by:
2LKDEADLOCKTHR  Thread "Default Executor-thread-1813" (0x00000000025B4E00)
3LKDEADLOCKWTR    which is waiting for:
4LKDEADLOCKMON      sys_mon_t:0x00000288F9BE4718 infl_mon_t: 0x00000288F9BE4798:
4LKDEADLOCKOBJ      org/apache/cxf/jaxrs/interceptor/CachedTime@0x00000000C5DD9548
3LKDEADLOCKOWN    which is owned by:
2LKDEADLOCKTHR  Thread "Default Executor-thread-1733" (0x0000000002149400)
NULL

**Comment by jhanders34** [comment]
Looking at the code, this seems to be a problem with Java and not a problem with the liberty code.  The liberty code is doing the right thing for synchronization to not have a deadlock.  There is a newer version of OpenJDK + OpenJ9 V11 available.  Can you try that version and see if this problem still exists?  https://adoptopenjdk.net/?variant=openjdk11&jvmVariant=openj9

**Comment by c-koell** [comment]
I will try it .. but as described if i use 20.0.0.1 we see see no deadlock. After upgrading to 20.0.0.2 the deadlock occurs.
The Java Version is with both versions the same ...

**Comment by c-koell** [comment]
I have upgraded to openjdk version 11.0.7 2020-04-14.
At the moment we see no more deadlocks. I will observe it the today and inform you about the status.
For me it's a little bit confusing how a deadlock can occur on a newer openliberty version, if the main problem seems to be in the jdk itself.

**Comment by jhanders34** [comment]
The code where you see a deadlock has not changed in over 2 years.  That is why I was concerned in thinking it was a Liberty problem.  With a new release the code is different all around which will cause the JDK to do different things with the code for JIT / AOT.  My guess is there is a pattern in the newer code that causes the OpenJ9 code to incorrectly think things are locked when they aren't or a lock is not correctly getting released or something.  Likely a defect against the OpenJ9 github would be in order to track it down:  https://github.com/eclipse/openj9/

**Comment by c-koell** [comment]
Thanks for the help i will close this issue due no more deadlock has occurred

**Comment by c-koell** [comment]
Hi @jhanders34 .

We have now a very similar problem in our environment again. The Rest Endpoint blocks new incoming requests in the same code as we had before.

I have attached two javacore files
[javacore.20211117.114259.7124.0001.txt](https://github.com/OpenLiberty/open-liberty/files/7554336/javacore.20211117.114259.7124.0001.txt)
[javacore.20211117.114302.7124.0002.txt](https://github.com/OpenLiberty/open-liberty/files/7554337/javacore.20211117.114302.7124.0002.txt)

As i can see we are again in the code of org/apache/cxf/jaxrs/interceptor/CachedTime

Java-Version: openjdk version "11.0.12" 2021-07-20 / IBM Semeru Runtime Open Edition 11.0.12.0 (build 11.0.12+7)
OpenLiberty: Open Liberty 21.0.0.10

Thanks for any help you can provide !

**Comment by c-koell** [comment]
@jhanders34 can you provide a link to the org.apache.cxf.jaxrs.interceptor.CachedTime what is used in openliiberty ?

I would like to have a look at the code to understand the deadlock mayby a bit more .. thanks !

**Comment by jhanders34** [comment]
> @jhanders34 can you provide a link to the org.apache.cxf.jaxrs.interceptor.CachedTime what is used in openliiberty ?
> 
> I would like to have a look at the code to understand the deadlock mayby a bit more .. thanks !

https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/interceptor/CachedTime.java

is the code.  Sorry for the delay.  I was on vacation last week.  Looking at your javacores, again I don't see how the code itself is causing the lock to remain locked by that thread since it shows that the code has moved on from CachedTime.  It is as if the Java code has decided to not release the lock.

**Comment by jhanders34** [comment]
You can try the 11.0.13 release to see if they fixed the problem there or not.

**Comment by c-koell** [comment]
Hi @jhanders34 ! Thanks for the link ! 
This time the lock occured in our production environment. We have deployed a new version of our Application. With that version also the OpenLiberty Version was changed from 21.0.0.3 to 21.0.0.10.
As a fast try we downgraded to OpenLiberty 21.0.0.3 and afterwards our application worked fine again.
This is the mysterium what i don't understand. Only with a Open Liberty Version update we get a deadlock ....

A problem for us is the bug https://github.com/eclipse-openj9/openj9/issues/13759
We have figured out that the sapjco.jar (SAP Library) checks the StackFrame and if <unknown class> comes as the package name the library does not work anymore. So we must wait for openjdk 11.0.14 with OpenJ9 Release 0.30.

I have created also a OpenJ9 Bug https://github.com/eclipse-openj9/openj9/issues/14037

I hope that we find a solution here. It is really anoying to update OpenLiberty or Java and you don't know if all works fine afterwards :-(

Thanks for any help
Claus
